dref/serializers.py

```diff
@@ -0,0 +1,217 @@
+    modified_by_details = UserNameSerializer(source='modified_by', read_only=True)
+    event_map_details = DrefFileSerializer(source='event_map', read_only=True)
+    images_details = DrefFileSerializer(source='images', many=True, read_only=True)
+    field_report_details = MiniFieldReportSerializer(source='field_report', read_only=True)
+    created_by_details = UserNameSerializer(source='created_by', read_only=True)
+    users_details = UserNameSerializer(source='users', many=True, read_only=True)
+    budget_file_details = DrefFileSerializer(source='budget_file', read_only=True)
+    cover_image_details = DrefFileSerializer(source='cover_image', read_only=True)
+    disaster_type_details = DisasterTypeSerializer(source='disaster_type', read_only=True)
+
+    class Meta:
+        model = Dref
+        fields = '__all__'
+        read_only_fields = ('modified_by', 'created_by', 'budget_file_preview')
+
+    def to_representation(self, instance):
+        data = super().to_representation(instance)
+        disability_people_per = data.get('disability_people_per', '')
+        if disability_people_per and len(disability_people_per.split('.')[0]) == 3:
+            data['disability_people_per'] = disability_people_per.split('.')[0]
+        people_per_urban = data.get('people_per_urban', '')
+        people_per_local = data.get('people_per_local', '')
+        if people_per_urban and len(people_per_urban.split('.')[0]) == 3:
+            data['people_per_urban'] = people_per_urban.split('.')[0]
+        if people_per_local and len(people_per_local.split('.')[0]) == 3:
+            data['people_per_local'] = people_per_local.split('.')[0]
+        return data
+
+    def validate(self, data):
+        event_date = data.get('event_date', None)
+        if event_date and data['type_of_onset'] not in [Dref.OnsetType.SLOW, Dref.OnsetType.SUDDEN]:
+            raise serializers.ValidationError({
+                'event_date': ugettext('Cannot add event_date if onset type not in {} or {}')
+                .format(Dref.OnsetType.SLOW.label, Dref.OnsetType.SUDDEN.label)
+            })
+        return data
+
+    def validate_images(self, images):
+        # Validate if images are provided
+        if images:
+            # Fetch current images
+            current_images_map = {
+                image.pk: image.file
+                for image in (self.instance and self.instance.images.all()) or []
+            }
+            for image in images:
+                # Ignore for current images (if file is not changed)
+                if image.id is not None and image.id in current_images_map and image.file == current_images_map[image.id]:
+                    continue
+                elif image.created_by != self.context['request'].user:
+                    raise serializers.ValidationError(
+                        ugettext('Only image owner can attach image(id: %s)' % image.id)
+                    )
```

Comment:
- Can we update DrefFile? I would recommend removing this action. Basically, each DrefFile represents a file, so only create/delete action makes sense.
- So, if we can't update DrefFile, then we can just do something like this to check if there are any images to which the user doesn't have access. `user .values_list('id', flat=True)` if we need to send id back as well)
```python3
images_id = [image.id for image in images]
Dref.objects.filter(
    # Not created by current user
    ~models.Q(created_by=self.context['request']),
    # Look into provided images
    id__in=images_id
).exclude(
    # Exclude already attached images
    id__in=self.instance.images.all()
).exists()
```